tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetSetService.scala

```diff
@@ -17,4 +19,29 @@ class PublicAssetSetService @Inject()(val identityRepository: IdentityRepository
         case Seq() => SMono.just(identityIds)
         case noExistIdentityIds => SMono.error(PublicAssetIdentityIdNotFoundException(noExistIdentityIds))
       }
-}
\ No newline at end of file
+
+  def create(username: Username, creationRequest: PublicAssetCreationRequest): SMono[PublicAssetStorage] =
+    SMono(publicAssetRepository.create(username, creationRequest))
+      .onErrorResume {
+        case _: PublicAssetQuotaLimitExceededException => cleanUpPublicAsset(username)
+          .filter(cleanedUpSize => cleanedUpSize >= creationRequest.size.value)
+          .switchIfEmpty(SMono.error(new PublicAssetQuotaLimitExceededException))
+          .flatMap(_ => SMono(publicAssetRepository.create(username, creationRequest)))
+      }
+
+  // return the total size of public assets that was cleaned up
+  def cleanUpPublicAsset(username: Username): SMono[Long] =
+    SFlux(identityRepository.list(username))
+      .map(_.id)
+      .collectSeq()
+      .flatMapMany(existIdentityIds => listPublicAssetNeedToDelete(username, existIdentityIds))
+      .concatMap(publicAssetMetadata => SMono(publicAssetRepository.remove(username, publicAssetMetadata.id))
+        .`then`(SMono.just(publicAssetMetadata.size.value)), 1)
+      .collectSeq()
+      .map(_.sum)
+
+  def listPublicAssetNeedToDelete(username: Username, existIdentityIds: Seq[IdentityId]): SFlux[PublicAssetMetadata] =
+    SFlux(publicAssetRepository.listPublicAssetMetaData(username))
+      .filter(publicAssetMetadata => publicAssetMetadata.identityIds.isEmpty || !publicAssetMetadata.identityIds.exists(identityId => existIdentityIds.contains(identityId)))
```

Comment:
Simpler:

```
!publicAssetMetadata.identityIds.exists(identityId => existIdentityIds.contains(identityId))
```

The empty clause is implicitly handled in the following condition.